### PR TITLE
Slicer and an unslicer

### PR DESCRIPTION
**Pull request by JorisCos**
### About this PR
This PR introduce a slicer and an unslicer to slice signals, feed the sliced signals to a model and reconstruct the unsliced signal afterwards. This could be useful for very long signals that may cause code to break and also allows models (like SEGAN) that take a precise input length to handle longer signals.

### Proposed slicer
The proposed slicer uses hanning window and can handle overlap between 0 to 50 %. I did not find an elegant way to implement a slicer where more than two chunks are overlapping at the same time. Small numerical errors are introduced during the reconstruction phase they are around 1E-7.

### CODE REVIEW

File: tests/utils/torch_utils_test.py [hunk 1]
```diff
@@ -36,9 +36,11 @@ def test_loader_module():
 def test_loader_submodule():
     class SuperModule(nn.Module):
         """ nn.Module subclass that holds a model under self.whoever """
+
```
Comment (mpariente) [inline tests/utils/torch_utils_test.py:39]: No need to rewrite parts of the code that the PR doesn't address.

File: asteroid/utils/torch_utils.py [hunk 2]
```diff
@@ -123,3 +123,107 @@ def are_models_equal(model1, model2):
...
+                                                    * w_up
+    # Cut unwanted padding
+    reconstructed = reconstructed[:, :, :original_shape[-1]]
+    return reconstructed
+
+import torch
+x = torch.randn(10, 1, 16384)
+sliced = slice(x)
+unsliced = unslice(sliced, x.shape)
+assert torch.mean(x-unsliced) < 1E-6
```
Comment (mpariente) [inline asteroid/utils/torch_utils.py:229]: Need new line please.

Comment (mpariente) [inline asteroid/utils/torch_utils.py:229]: This should be removed right?

File: tests/utils/torch_utils_test.py [hunk 3]
```diff
@@ -53,3 +55,13 @@ def __init__(self, sub_model):
...
     model_2 = torch_utils.load_state_dict_in(state_dict, model_2)
     assert torch_utils.are_models_equal(model, model_2)
+
+
+def test_slicer():
+    x = torch.randn(10, 2, 16384)
+    sliced = torch_utils.slice(x)
+    unsliced = torch_utils.unslice(sliced, x.shape)
+    result = False
+    if torch.mean(x - unsliced) < 1E-6:
```
Comment (mpariente) [inline tests/utils/torch_utils_test.py:65]:
```python
from torch.testing import assert_allclose
assert_allclose(x, unsliced, atol=..., rtol=...)
```

File: asteroid/utils/torch_utils.py [hunk 4]
```diff
@@ -123,3 +123,107 @@ def are_models_equal(model1, model2):
...
     return True
+
+
+def slice(sources, w_size=16384, h_size=16384 // 2):
+    """  Slice a signal in slices of size w_size
+
+     Args:
+         sources (torch.Tensor): sources of shape [nb_sources,channels,samples]
+         w_size (int): size of the window
+         h_size (int): size of the stride
```
Comment (mpariente) [inline asteroid/utils/torch_utils.py:134]: I would call this stride though.

File: tests/utils/torch_utils_test.py [hunk 5]
```diff
@@ -53,3 +55,13 @@ def __init__(self, sub_model):
     # Apply our workaround torch_utils.load_state_dict_in and assert True.
     model_2 = torch_utils.load_state_dict_in(state_dict, model_2)
     assert torch_utils.are_models_equal(model, model_2)
+
+
+def test_slicer():
+    x = torch.randn(10, 2, 16384)
```
Comment (mpariente) [inline tests/utils/torch_utils_test.py:61]: Hmm, this doesn't need to be sliced no? 
So it actually doesn't unit test, or I'm missing something? 
Can we test with other shapes

File: asteroid/utils/torch_utils.py [hunk 6]
```diff
@@ -123,3 +123,107 @@ def are_models_equal(model1, model2):
...
+    if len_s < w_size:
+        new_size = list(sources.size()[:-1]) + [w_size]
+        padded = torch.zeros(new_size)
+        padded[:, :, :len_s] = sources
+        return padded.unsqueeze(1)
+    # Else source equal window just unsqueeze
+    return sources.unsqueeze(1)
+
+
+def unslice(sources, original_shape, w_size=16384, h_size=16384 // 2):
```
Comment (mpariente) [inline asteroid/utils/torch_utils.py:176]: Because we pad in the slicing, we should have a target_len argument to remove the zeros we added at the end right?

File: asteroid/utils/torch_utils.py [hunk 7]
```diff
@@ -123,3 +123,107 @@ def are_models_equal(model1, model2):
...
+
+def slice(sources, w_size=16384, h_size=16384 // 2):
+    """  Slice a signal in slices of size w_size
+
+     Args:
+         sources (torch.Tensor): sources of shape [nb_sources,channels,samples]
+         w_size (int): size of the window
+         h_size (int): size of the stride
+     Returns:
+         sliced signal (torch.Tensor): shape [nb_sources,nb_slices,channels,
```
Comment (mpariente) [inline asteroid/utils/torch_utils.py:136]: At inference, we are expecting signals with shape (1, 1, wal_len), (1, wav_len) or (wav_len,) but we might not want to make the assumption the shape is always (*, *, time).

Comment (mpariente) [reply asteroid/utils/torch_utils.py:136]: A recursive solution would work. Or a class interface that handles the reshaping.

Comment (mpariente) [reply asteroid/utils/torch_utils.py:136]: If we want to support several input shapes, their inversion should be unique (unslice(slice(x)) should equal x).

File: asteroid/utils/torch_utils.py (hunk 2 above)
Comment (JorisCos) [reply asteroid/utils/torch_utils.py:229]: Yes

File: asteroid/utils/torch_utils.py (hunk 6 above)
Comment (JorisCos) [reply asteroid/utils/torch_utils.py:176]: Yes, that's what original_shape arg in unslice is. However, we can't remove the padding in the output of `slice() `otherwise the slicing method is unsuitable for models that only take a specific shape as input.

File: asteroid/utils/torch_utils.py (hunk 7 above)
Comment (JorisCos) [reply asteroid/utils/torch_utils.py:136]: I think i'm going to try the class interface.

File: tests/utils/torch_utils_test.py (hunk 5 above)
Comment (JorisCos) [reply tests/utils/torch_utils_test.py:61]: No you are right

File: asteroid/utils/torch_utils.py (hunk 6 above)
Comment (mpariente) [reply asteroid/utils/torch_utils.py:176]: Oh yes, sorry. 
Then this should be a kwarg with None as default. 
If None, we don't remove any padding.

**Comment by mpariente** [comment]
Closing in favor of #191
